yew/src/html/listener/listener_web_sys.rs

```diff
@@ -1,100 +1,273 @@
 // Inspired by: http://package.elm-lang.org/packages/elm-lang/html/2.0.0/Html-Events
+
+use wasm_bindgen::JsCast;
+use web_sys::{Element, Event, HtmlInputElement, HtmlSelectElement, HtmlTextAreaElement};
+
+macro_rules! impl_action {
+    ($($action:ident($type:ident) -> $ret:path => $convert:path)*) => {$(
+        impl_action!($action($type, ::std::u8::MIN) -> $ret => $convert);
+    )*};
+    ($($action:ident($type:ident, $flags:path) -> $ret:path => $convert:path)*) => {$(
+        /// An abstract implementation of a listener.
+        #[doc(hidden)]
+        pub mod $action {
+            use crate::callback::Callback;
+            use crate::virtual_dom::Listener;
+
+            /// A wrapper for a callback which attaches event listeners to elements.
+            #[derive(Clone, Debug)]
+            pub struct Wrapper {
+                callback: Callback<Event>,
+            }
+
+            impl Wrapper {
+                /// Create a wrapper for an event-typed callback
+                pub fn new(callback: Callback<Event>) -> Self {
+                    Wrapper { callback }
+                }
+            }
+
+            /// And event type which keeps the returned type.
+            pub type Event = $ret;
+
+            impl Listener for Wrapper {
+                fn kind(&self) -> &'static str {
+                    stringify!($action)
+                }
+
+                fn handle(&self, event: web_sys::Event) {
+                    self.callback.emit($convert(event));
+                }
+
+                fn flags(&self) -> u8 {
+                    match &self.callback {
+                        Callback::Callback{flags, ..} => (*flags).unwrap_or($flags),
+                        _ => $flags,
+                    }
+                }
+            }
+        }
+    )*};
+}
+
+pub(crate) fn nop_convert<T>(t: T) -> T {
+    t
+}
+
+pub(crate) fn cast_event<T>(e: web_sys::Event) -> T
+where
+    T: wasm_bindgen::JsCast,
+{
+    e.dyn_into().unwrap()
+}
+
+// Reduces repetition for common cases
+macro_rules! impl_short {
+    ($($action:ident)*) => {
+        impl_action! {
+            $(
+                $action(Event) -> web_sys::Event => crate::html::listener::nop_convert
+            )*
+        }
+    };
+    ($($action:ident($type:ident))*) => {
+        impl_action! {
+            $(
+                $action($type) -> web_sys::$type  => crate::html::listener::cast_event
+            )*
+        }
+    };
+}
+
+// Unspecialized event type
+impl_short! {
```

Comment:
I love how you cut down the boilerplate here. One problem I have is the removal of the deliberate formatting.
The structure was there so it's easier to tell which types are part of the living standard and which are unofficial or experimental. Could you try to bring that back a bit?